specs/phase0/p2p-interface.md

```diff
@@ -746,10 +748,25 @@ The request MUST be encoded as an SSZ-container.
 The response MUST consist of zero or more `response_chunk`.
 Each _successful_ `response_chunk` MUST contain a single `SignedBeaconBlock` payload.
 
-Clients MUST keep a record of signed blocks seen since the start of the weak subjectivity period
-and MUST support serving requests of blocks up to their own `head_block_root`.
+Clients MUST keep a record of signed blocks seen on the epoch range
+`[max(GENESIS_EPOCH, current_epoch - MIN_EPOCHS_FOR_BLOCK_REQUESTS), current_epoch]`
+where `current_epoch` is defined by the current wall-clock time,
+and clients MUST support serving requests of blocks on this range.
+
+Synced clients unable to reply to Block requests within the
+`MIN_EPOCHS_FOR_BLOCK_REQUESTS` epoch range MAY get descored or disconnected at any time.
+Note, due to this it is risky behaviour to begin participating as a full node at the head if having
+not yet backfilled on this range.
```

Comment:
Currently, teku's checkpoint sync will load a provided finalized checkpoint state, forward sync to catch up to head and get the node online quickly, then backfill blocks requesting them in reverse order to ensure that we form a hash chain and download the correct history.

To accommodate the suggestions here, we'd need to instead calculate the oldest required slot, forward sync from there, and hope that we sync the correct chain that matches up with the provided checkpoint state.  If we end up downloading a malicious chain, we'd have to throw away the downloaded blocks and try again.  

This seems suboptimal.  In the first strategy, we're guaranteed to sync the correct blocks and get online quickly which is very helpful to end users.  In the second strategy, we're not guaranteed to find the correct chain, and we're forcing users to wait much longer than necessary to get their nodes online.

I see the point that we don't want to lose old blocks.  Long term, if we stick with the idea of updating `Status` to advertise the oldest available block, it will be simple enough to enforce this:  we just have to spot check blocks in the advertised range and check that the advertised range is being backfilled over time (the oldest available block advertised in `Status` is being updated at a reasonable rate).  

If we're really concerned about the possibility of clients not backfilling in the short-term, we could do a binary search when nodes connect to find their oldest available block.  And check every so often that older blocks have become available (similar to the long-term strategy, but slightly less efficient). 

The ability to reliably get up and running quickly is extremely useful for users and it's one of the nicer features of eth2 in my opinion.  I don't think we should be discouraging this.

Reply:
This does not suggest doing a forward sync. It just suggests that if your peers notice you haven't backfilled, they can disconnect.

And it implies that your peers might check and MAY disconnect (e.g. could use random sample or binary search). [I even had "nodes MAY submit random queries on that range and MAY disconnect upon failure to respond" but removed it]

Pretty much what it says is that once you start participating a a full node at the head, you are subject to having blocks on that range. Again, *it does not* say you must forward sync.

Something I wrote must be unclear. Can you re-read now thatt you know my intention and suggest how I might clarify?

Reply:
I tried to highlight these lines but I guess it didn't come across:
```
Note, due to this it is risky behaviour to begin participating as a full node at the head if having
not yet backfilled on this range.
```

To me that sounds like we're discouraging nodes from advertising a recent head block (via `Status`) without first having backfilled the historical blocks.  And it seems to me that way to avoid this is to "backfill" in order starting from the oldest block in the required range up to the checkpoint.  This way, we always have the blocks from the beginning of the range up the head block that we're advertising (the latest backfilled block).

Maybe I'm misunderstanding what it means to "begin participating as a full node at the head"?  Otherwise, it might be worth talking more about what strategies nodes may take while they're in the process of backfilling.

Reply:
I also thought that having an explicit `UNAVAILABLE` error which could be used while backfilling made it clearer how nodes should behave while in the process of backfilling.

Reply:
> And it seems to me that way to avoid this is to "backfill" in order starting from the oldest block in the required range up to the checkpoint. 

That is certainly one way, but the spec does not say which way you should back-fill, it just says that you SHOULD. And if you don't and people ask you for blocks that you don't have, you might be subject to negative consequences

You're right in that the `Status` is used means that you must send it even before backfilling (because you connected to peers) and thus it isn't clear what should be included in it during backfill. Even if you back-fill from back to front, you still know of a more recent finalized head/root/slot

I'm trying to decide the simplest way to say, "I'm not ready to serve" without modifying the current `Status` data structure. Any ideas?

Reply:
> hope that we sync the correct chain that matches up with the provided checkpoint state

this is actually a valuable sanity check that the provided checkpoint state is sane - if it _doesn't_ match up (in terms of block hashes), something is wrong: the checkpoint state, the nodes around you etc and the beacon node cannot participate. It's not so much a hope, as a hard requirement.

> much longer

Syncing pre-checkpoint blocks is much faster than syncing "normal" blocks - there's no need to do state transitions and finality processing: it needs signature and tree root verification only.

> If we end up downloading a malicious chain, we'd have to throw away the downloaded blocks and try again.

This should be noticeable fairly quickly - ie if you're downloading from several peers, a chain must be formed by the received blocks - they also must have valid signatures from the validator set in the checkpoint state - `SLOTS_PER_HISTORICAL_ROOT` block roots are also available in the checkpoint state for early hash verification.

> `UNAVAILABLE`

The spec can't at the same time _require_ that clients make blocks available and give validity to not doing so through an error code - it must choose: either it's ok to drop blocks / not backfill _or_ clients that don't provide blocks can be sanctioned - the main sanction we have available is to disconnect or descore such clients - they are an undue burden upon the network as a whole, which is what this PR points out.

> I'm trying to decide the simplest way to say, "I'm not ready to serve" without modifying the current Status data structure. Any ideas?

set head block to `block_at_slot(now - MIN_EPOCHS_FOR_BLOCK_REQUESTS)` and advance it as backfilling advances - this tells others that you're ready or not for syncing up to the given block/slot - it also means that you should be able to reply to block requests for the range `[now - MIN_EPOCHS_FOR_BLOCK_REQUESTS, Status.head_slot)`. `Status.finalized_epoch`  is a bit more tricky: it can follow `head_slot` up to the subjectivity checkpoint most likely, though this is not *strictly* true - one could also set it to the WS checkpoint from the start, though this is a bit weird (how can head be < than finalized?)

before `block_at_slot(now - MIN_EPOCHS_FOR_BLOCK_REQUESTS)` is known, it seems reasonable to set it to the genesis block - it also seems reasonable to consider receiving a block signed by the "correct" validator as a good "head" block once received from the first range request.

An interesting question of course what to do with clients that are "not ready to serve" yet are gossipping attestations that point to blocks newer than their status - they are either cheating in their `Status` or are not validating attestations as they should - this provides another data point by which to descore leeching clients.

Reply:
> I'm trying to decide the simplest way to say, "I'm not ready to serve" without modifying the current `Status` data structure. Any ideas?

We could set the head root / slot to a special value (zero?).  And leave the finalized root / slot pointing to the WS checkpoint.

Reply:
> The spec can't at the same time _require_ that clients make blocks available and give validity to not doing so through an error code - it must choose: either it's ok to drop blocks / not backfill _or_ clients that don't provide blocks can be sanctioned - the main sanction we have available is to disconnect or descore such clients - they are an undue burden upon the network as a whole, which is what this PR points out.

I don't think an error code is mutually exclusive w/ requiring clients to backfill.  It's just a recognition of the fact that backfilling takes time.  If a node always returns this error code and never backfills, then it can be considered malicious.  There's also nothing preventing nodes from descoring / disconnecting a peer that returns this error.  I just think that we don't want to mandate super strict behaviour here that prevents clients from implementing a reverse backfill strategy.

Reply:
To add to what Meredith said, I think we're undervaluing the benefit and importance of doing snapshot syncs and being overly concerned with the risk of clients not backfilling blocks. With the weak subjectivity period the most secure way to sync the chain is via snapshot sync - anything else leaves you at least vulnerable to DOS attacks by feeding you the wrong chain and making you process it all the way up to the weak subjectivity checkpoint at which point you throw out all that work and start again.  Snapshot sync is also a huge win for users who are able to get a node up and validating in under a minute regardless of the length of the chain.

Then in terms of the risk of not backfilling, the spec is only requiring backfill to the start of the weak subjectivity period, so syncing from genesis may become impossible anyway (though Teku is backfilling all the way to genesis at the moment even after a snapshot sync since genesis sync is still so common).  If you are syncing, then naturally you would disconnect any peers that can't provide the blocks you need - that may be because they didn't backfill but more likely because they themselves are still syncing or are on a different fork. That's no different to disconnecting peers that aren't on the attestation subnet you need - they aren't malicious, just useless to you.

So I'm really confused as to why we're specifically adding to the spec that it's risky to take full advantage of snapshot sync when its got such clear benefits and I'm struggling to see what the downside is of not being able to identify nodes that haven't completed backfill yet.  Even the threat of "leeching" clients that don't backfill at all seems pretty limited - they're going to make up quite a small part of the network and will naturally wind up building up historic blocks as time progresses and they import them.  Nodes that are already in sync won't notice they didn't backfill and new nodes would be starting from a more recent snapshot anyway so probably also wouldn't notice.

I'd understand more if the requirement was to backfill all blocks to genesis but currently I don't understand what threat we're trying to mitigate here.

Reply:
> I'd understand more if the requirement was to backfill all blocks to genesis but currently I don't understand what threat we're trying to mitigate here.

the objective would to ensure that regardless where in the WS period you are, you have blocks in the network to sync from - one could calculate a more accurate period but what this PR does it simply take the maximum reasonable WS period and sync from there - resyncing the blocks also acts as a trivial guard against malicious WS checkpoint services.

the other objective is to make a small adjustment that fits within the existing protocol for HF1 and then make a bigger change later on, when there's some more meaningful changes around to care about.

> risky to take full advantage of snapshot sync

you're still taking advantage of snapshot sync: you quickly download a few extra blocks, check their sigs and hashes, and you're good to go.

> Even the threat of "leeching" clients

Judging by the scarcity of LES servers on eth1, it's not a theoretical problem. Part of your job as a validator is to supply a copy of the blocks for a while - at least, that is the case for now in phase0.

I see this PR mainly as a stepping stone that we can improve upon later by relaxing the rules further where it's safe to do so. As mentioned in other comments, once clients start supplying different ranges of blocks, in order to sanely be able to find viable peers, one needs an improved discovery mechanism at least.

Reply:
> the objective would to ensure that regardless where in the WS period you are, you have blocks in the network to sync from - one could calculate a more accurate period but what this PR does it simply take the maximum reasonable WS period and sync from there - resyncing the blocks also acts as a trivial guard against malicious WS checkpoint services.

Given that the majority of nodes are not freshly sync'd this seems like something that will be pretty straight forward even without actively backfilling blocks.  And I'm suggesting we *do* actively backfill blocks just that we shouldn't have to arbitrarily wait for that process to complete before starting to attest or participating fully in gossip etc at the chain head.

Resyncing the blocks back to genesis provides a guard against malicious checkpoint services, but syncing back to the start of the WS period doesn't provide any extra verification because it doesn't attach to any known state.

> the other objective is to make a small adjustment that fits within the existing protocol for HF1 and then make a bigger change later on, when there's some more meaningful changes around to care about.

The simplest change we could make is to just define an error response code that indicates the node doesn't have that block.  This change doesn't actually solve the problem because any node syncing from genesis still has no way to know if a remote peer has blocks back to genesis or only back to the weak subjectivity period, so still can't differentiate between a claim that "there were no blocks" or "I don't have those blocks".

> > risky to take full advantage of snapshot sync
> 
> you're still taking advantage of snapshot sync: you quickly download a few extra blocks, check their sigs and hashes, and you're good to go.

But we've added a whole bunch of complexity about how to behave when you're in sync but not backfilled and having to track where the backfill process is up to etc.  But we don't get any actual benefit from that complexity.

> > Even the threat of "leeching" clients
> 
> Judging by the scarcity of LES servers on eth1, it's not a theoretical problem. Part of your job as a validator is to supply a copy of the blocks for a while - at least, that is the case for now in phase0.

The equivalent in Eth1 is not the LES protocol, but the direct equivalent of being able to request blocks from peers. Not all nodes on Eth1 have backfilled all blocks and there's no requirement to have all historic blocks before participating fully in the network.  If a node needs specific blocks and one of it's peers doesn't have them, it is allowed to disconnect because the peer is useless to them.

> I see this PR mainly as a stepping stone that we can improve upon later by relaxing the rules further where it's safe to do so. 

What is it a stepping stone towards though? What problem are we trying to actually solve here? We've seen both Eth1 and Eth2 function well without anything complicated around ensuring peers backfill blocks.

> As mentioned in other comments, once clients start supplying different ranges of blocks, in order to sanely be able to find viable peers, one needs an improved discovery mechanism at least.

No such mechanism exists on Eth1 and its never been an issue.  I find it very hard to believe it will be a problem on the beacon chain as that would imply that the vast majority of nodes are very recently sync'd which seems extremely unlikely.

Reply:
So I agree with @arnetheduck that it is risky to *not* define this behavior has *potentially* malicious or lazy. If a peer can consistently say they don't have any blocks to serve, then it becomes much easier for nodes to not participate in block sync at all.

A compromise here might be something like "If a peer repeatedly does not have blocks available to serve since the start of the weak subjectivity period, they can be descored or disconnected". We then might be a bit more careful on defining "repeatedly" -- e.g. "After initial connection, if for more than 10 minutes a peer does not have..." 

Just an example. I know this adds complexity *if* a node wants to penalize such behavior but leaves open the option to do so

Reply:
> A compromise here might be something like "If a peer repeatedly does not have blocks available to serve since the start of the weak subjectivity period, they can be descored or disconnected". We then might be a bit more careful on defining "repeatedly" -- e.g. "After initial connection, if for more than 10 minutes a peer does not have..."

That sounds reasonable.  I think it's also reasonable to say that if a node can't return blocks that your node /needs/ then it should disconnect immediately, possibly with a short term ban. If you're trying to sync and the remote node is trying to backfill you aren't useful to each other and are better of not being connected to each other.

That said, with the network conditions we currently have and experience with block availability on eth1, I'd say that actively requesting blocks you don't actually need is an unwarranted burden on the network.  We may need to perform those kinds of checks in the future if we can't find a better solution, but given the bandwidth concerns users already have adding more load isn't ideal.